storage/deposit/merkle/node.go

```diff
@@ -0,0 +1,314 @@
+}
+
+// GetRoot returns the root of the Merkle tree.
+func (l *LeafNode) GetRoot() [32]byte {
+	return l.hash
+}
+
+// IsFull returns whether there is space left for deposits.
+// A LeafNode will always return true as it is the last node
+// in the tree and therefore can't have any deposits added to it.
+func (*LeafNode) IsFull() bool {
+	return true
+}
+
+// Finalize marks deposits of the Merkle tree as finalized.
+func (l *LeafNode) Finalize(_, _ uint64) (TreeNode, error) {
+	return &FinalizedNode{1, l.hash}, nil
+}
+
+// GetFinalized returns a list of hashes of all the finalized nodes and the
+// number of deposits.
+func (*LeafNode) GetFinalized(result [][32]byte) (uint64, [][32]byte) {
+	return 0, result
+}
+
+// PushLeaf adds a new leaf node at the next available zero node.
+func (*LeafNode) PushLeaf([32]byte, uint64) (TreeNode, error) {
+	return nil, ErrLeafNodeCannotPushLeaf
+}
+
+// Right returns nil as a leaf node is the last node and can't have any
+// children.
+func (*LeafNode) Right() TreeNode {
+	return nil
+}
+
+// Left returns nil as a leaf node is the last node and can't have any children.
+func (*LeafNode) Left() TreeNode {
+	return nil
+}
+
+// Equals returns true if this leaf node is equal to the other node.
+func (l *LeafNode) Equals(other TreeNode) bool {
+	if l == nil && other == nil {
+		return true
+	}
+	if l == nil || other == nil {
+		return false
+	}
+	ln, ok := other.(*LeafNode)
+	if !ok {
+		return false
+	}
+	return l.hash == ln.hash
+}
+
+// InnerNode represents an inner node with two children and satisfies the
+// TreeNode interface.
+type InnerNode struct {
+	left, right TreeNode
+	hasher      merkle.Hasher[[32]byte]
+}
+
+// GetRoot returns the root of the Merkle tree.
+func (n *InnerNode) GetRoot() [32]byte {
+	left := n.left.GetRoot()
+	right := n.right.GetRoot()
+	return n.hasher.Combi(left, right)
+}
+
+// IsFull returns whether there is space left for deposits.
+func (n *InnerNode) IsFull() bool {
+	return n.right.IsFull()
+}
+
+// Finalize marks deposits of the Merkle tree as finalized.
+func (n *InnerNode) Finalize(
+	depositsToFinalize uint64,
+	depth uint64,
+) (TreeNode, error) {
+	var err error
+	deposits := pow.TwoToThePowerOf(depth)
+	if deposits <= depositsToFinalize {
+		return &FinalizedNode{deposits, n.GetRoot()}, nil
+	}
+	if depth == 0 {
+		return nil, ErrZeroDepth
+	}
+	n.left, err = n.left.Finalize(depositsToFinalize, depth-1)
+	if err != nil {
+		return nil, err
+	}
+
+	//nolint:mnd // spec.
+	if depositsToFinalize > deposits/2 {
+		remaining := depositsToFinalize - deposits/2
+		n.right, err = n.right.Finalize(remaining, depth-1)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return n, nil
+}
+
+// GetFinalized returns a list of hashes of all the finalized nodes and the
+// number of deposits.
+func (n *InnerNode) GetFinalized(result [][32]byte) (uint64, [][32]byte) {
+	leftDeposits, result := n.left.GetFinalized(result)
+	rightDeposits, result := n.right.GetFinalized(result)
```

Comment:
this is reusing the `result` passed as param to `GetFinalized`, and not creating a new shadow variable?